sound/soc/intel/boards/tgl_rt1308.c

```diff
@@ -0,0 +1,190 @@
+	if (err < 0) {
+		dev_err(card->dev, "Failed to set RT1308 PLL: %d\n", err);
+		return err;
+	}
+
+	/* Set rt1308 sysclk */
+	err = snd_soc_dai_set_sysclk(codec_dai, RT1308_FS_SYS_S_PLL, pll_out,
+				     SND_SOC_CLOCK_IN);
+	if (err < 0) {
+		dev_err(card->dev, "Failed to set RT1308 SYSCLK: %d\n", err);
+		return err;
+	}
+
+	return 0;
+}
+
+/* machine stream operations */
+static struct snd_soc_ops tgl_rt1308_ops = {
+	.hw_params = tgl_rt1308_hw_params,
+};
+
+static const struct snd_soc_dapm_widget tgl_rt1308_dapm_widgets[] = {
+	SND_SOC_DAPM_SPK("Speakers", NULL),
+	SND_SOC_DAPM_MIC("SoC DMIC", NULL),
+};
+
+static const struct snd_kcontrol_new tgl_rt1308_controls[] = {
+	SOC_DAPM_PIN_SWITCH("Speakers"),
+};
+
+static const struct snd_soc_dapm_route tgl_rt1308_dapm_routes[] = {
+	{ "Speakers", NULL, "SPOL" },
+	{ "Speakers", NULL, "SPOR" },
+
+	/* digital mics */
+	{"DMic", NULL, "SoC DMIC"},
+};
+
+SND_SOC_DAILINK_DEF(ssp2_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("SSP2 Pin")));
+
+SND_SOC_DAILINK_DEF(ssp2_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC("i2c-10EC1308:00", "rt1308-aif")));
+
+SND_SOC_DAILINK_DEF(platform,
+	DAILINK_COMP_ARRAY(COMP_PLATFORM("0000:00:1f.3")));
+
+SND_SOC_DAILINK_DEF(dmic_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("DMIC01 Pin")));
+SND_SOC_DAILINK_DEF(dmic_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC("dmic-codec", "dmic-hifi")));
+SND_SOC_DAILINK_DEF(dmic16k,
+	DAILINK_COMP_ARRAY(COMP_CPU("DMIC16k Pin")));
+
+static struct snd_soc_dai_link tgl_rt1308_dailink[] = {
+	{
+		.name		= "SSP2-Codec",
+		.id		= 0,
+		.no_pcm		= 1,
+		.ops		= &tgl_rt1308_ops,
+		.dpcm_playback = 1,
+		.nonatomic = true,
+		SND_SOC_DAILINK_REG(ssp2_pin, ssp2_codec, platform),
+	},
+	{
+		.name = "dmic01",
+		.id = 1,
+		.ignore_suspend = 1,
+		.dpcm_capture = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(dmic_pin, dmic_codec, platform),
+	},
+	{
+		.name = "dmic16k",
+		.id = 2,
+		.ignore_suspend = 1,
+		.dpcm_capture = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(dmic16k, dmic_codec, platform),
+	},
+};
+
+/* audio machine driver */
+static struct snd_soc_card tgl_rt1308_card = {
+	.name         = "tgl_rt1308",
+	.owner        = THIS_MODULE,
+	.dai_link     = tgl_rt1308_dailink,
+	.num_links = ARRAY_SIZE(tgl_rt1308_dailink),
+	.controls = tgl_rt1308_controls,
+	.num_controls = ARRAY_SIZE(tgl_rt1308_controls),
+	.dapm_widgets = tgl_rt1308_dapm_widgets,
+	.num_dapm_widgets = ARRAY_SIZE(tgl_rt1308_dapm_widgets),
+	.dapm_routes = tgl_rt1308_dapm_routes,
+	.num_dapm_routes = ARRAY_SIZE(tgl_rt1308_dapm_routes),
+	.late_probe = tgl_card_late_probe,
+};
+
+static int tgl_rt1308_probe(struct platform_device *pdev)
+{
+	struct snd_soc_acpi_mach *mach;
+	struct tgl_card_private *ctx;
+	struct snd_soc_card *card = &tgl_rt1308_card;
+	int ret;
+
+	card->dev = &pdev->dev;
+
+	ctx = devm_kzalloc(&pdev->dev, sizeof(*ctx), GFP_KERNEL);
+	if (!ctx)
+		return -ENOMEM;
+
+	if (IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI))
+		INIT_LIST_HEAD(&ctx->hdmi_pcm_list);
```

Comment:
The HDMI support seems a bit partial. We don't support HDAC_HDMI with new mach drivers, so I recommend either dropping it (like this L162-163), or adding HDMI support like e.g. sof_pcm512x.c are other more recent machine drivers.